app/main.py

```diff
@@ -86,128 +88,56 @@
 }
 
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
-
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
-
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
-
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
+def apply_effects(knight: dict) -> None:
+    """Apply all effects (armor, weapon, potion) to a knight."""
+    # Initialize protection
+    knight["protection"] = 0
 
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
+    # Apply armor
+    for armor_piece in knight["armour"]:
+        knight["protection"] += armor_piece["protection"]
 
-    # mordred
-    mordred = knightsConfig["mordred"]
+    # Apply weapon
+    knight["power"] += knight["weapon"]["power"]
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+    # Apply potion if exists
+    if knight["potion"] is not None:
+        for stat, value in knight["potion"]["effect"].items():
+            if stat == "protection":
+                knight["protection"] += value
+            else:
+                knight[stat] += value
```

Comment:
The code correctly checks if a potion exists and applies its effects. However, it does not handle the case where the potion effect might reduce an attribute (e.g., hp) below zero. Consider adding a check to ensure that attributes like hp do not fall below zero after applying potion effects.